### PR TITLE
support sealer upgrade

### PR DESCRIPTION
**Pull request by wizardlyk**
### Describe what this PR does / why we need it
Support sealer upgrade

### Describe how you did it
Distribute materials by sealer & Upgrade cluster by plugin。

### Describe how to verify it
```
sealer run ack-agility-registry.cn-shanghai.cr.aliyuncs.com/ecp_builder/ackdistro:v1-22-15-ack-upgrade-lyk-test --mode upgrade
```

### CODE REVIEW

File: pkg/cluster-runtime/upgrader.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	extension, err := i.ImageEngine.GetSealerImageExtension(&common2.GetImageAnnoOptions{ImageNameOrID: clusterImageName})
+	if err != nil {
+		return fmt.Errorf("failed to get cluster image extension: %s", err)
+	}
+
+	if extension.Type != v12.KubeInstaller {
+		return fmt.Errorf("exit install process, wrong cluster image type: %s", extension.Type)
+	}
+
+	// set HostAlias
```
Comment (VinceCui) [inline pkg/cluster-runtime/upgrader.go:39]: no need set hostalias in upgrade

File: cmd/sealer/cmd/cluster/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+var upgradeFlags *types.UpgradeFlags
+
+var longUpgradeCmdDescription = `upgrade command is used to upgrade cluster.
+User can upgrade cluster by explicitly specifying host IP`
+
+var exampleForUpgradeCmd = `
+upgrade cluster:
+  sealer upgrade --masters 192.168.0.1 --nodes 192.168.0.2 -p Sealer123
+  sealer upgrade --masters 192.168.0.1-192.168.0.3 --nodes 192.168.0.4-192.168.0.6 -p Sealer123
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/cluster/upgrade.go:44]: could i ask why we set master ips ?  do node scale after upgrade ? 
i have a idea:  `sealer upgrade  ${cluster-name} --image ${wanted-version}`

File: cmd/sealer/cmd/cluster/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+	imagecommon "github.com/sealerio/sealer/pkg/define/options"
+	"github.com/sealerio/sealer/pkg/imagedistributor"
+	"github.com/sealerio/sealer/pkg/imageengine"
+	"github.com/sealerio/sealer/pkg/infradriver"
+	"sigs.k8s.io/yaml"
+)
+
+var upgradeFlags *types.UpgradeFlags
+
+var longUpgradeCmdDescription = `upgrade command is used to upgrade cluster.
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/cluster/upgrade.go:38]: if this cluster is not installed by sealer , can we upgrade it ?

File: cmd/sealer/cmd/cluster/upgrade.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+var exampleForUpgradeCmd = `
+upgrade cluster:
+  sealer upgrade --masters 192.168.0.1 --nodes 192.168.0.2 -p Sealer123
+  sealer upgrade --masters 192.168.0.1-192.168.0.3 --nodes 192.168.0.4-192.168.0.6 -p Sealer123
+`
+
+func NewUpgradeCmd() *cobra.Command {
+	upgradeCmd := &cobra.Command{
+		Use:     "upgrade",
+		Short:   "upgrade kubernetes version of cluster",
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/cluster/upgrade.go:50]: could we upgrade others are similar to k8s clusters  in this implementation , like k3s, k0s

File: cmd/sealer/cmd/cluster/upgrade.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+		return err
+	}
+
+	logrus.Infof("succeeded in upgrading cluster with image %s", clusterImageName)
+
+	return nil
+}
+
+func upgradeWithClusterfile(clusterFile string, upgradeFlags *types.UpgradeFlags) error {
+	return nil
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/cluster/upgrade.go:205]: Confused about this func 🤔

File: pkg/cluster-runtime/upgrader.go (hunk 1 above)
Comment (wizardlyk) [reply pkg/cluster-runtime/upgrader.go]: removed

File: cmd/sealer/cmd/cluster/upgrade.go (hunk 2 above)
Comment (wizardlyk) [reply cmd/sealer/cmd/cluster/upgrade.go]: updated

File: cmd/sealer/cmd/cluster/upgrade.go (hunk 3 above)
Comment (wizardlyk) [reply cmd/sealer/cmd/cluster/upgrade.go:33]: not supported currently

File: cmd/sealer/cmd/cluster/upgrade.go (hunk 5 above)
Comment (wizardlyk) [reply cmd/sealer/cmd/cluster/upgrade.go]: removed

File: cmd/sealer/cmd/cluster/upgrade.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cluster
+
+import (
+	"fmt"
+	"github.com/sealerio/sealer/cmd/sealer/cmd/types"
```
Comment (Starnop) [inline cmd/sealer/cmd/cluster/upgrade.go:19]: go import group

File: cmd/sealer/cmd/cluster/upgrade.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+)
+
+var upgradeFlags *types.UpgradeFlags
+
+var longUpgradeCmdDescription = `upgrade command is used to upgrade cluster.
+User can upgrade cluster by explicitly specifying image name`
+
+var exampleForUpgradeCmd = `
+upgrade cluster:
+  sealer upgrade --i upgrade:v1.22.15
```
Comment (Starnop) [inline cmd/sealer/cmd/cluster/upgrade.go:38]: s/--i/-i

File: cmd/sealer/cmd/cluster/upgrade.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+`
+
+func NewUpgradeCmd() *cobra.Command {
```
Comment (Starnop) [inline cmd/sealer/cmd/cluster/upgrade.go:41]: define upgrade cmd just like `sealer run`

@kakaZhou719 WDYT?

File: cmd/sealer/cmd/cluster/upgrade.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+	if err != nil {
+		return err
+	}
+
+	imageMounter, err := imagedistributor.NewImageMounter(imageEngine, clusterHostsPlatform)
+	if err != nil {
+		return err
+	}
+
+	imageMountInfo, err := imageMounter.Mount(clusterImageName)
```
Comment (Starnop) [inline cmd/sealer/cmd/cluster/upgrade.go:117]: maybe should use new cluster image here

File: pkg/cluster-runtime/hook.go [hunk 10]
```diff
@@ -55,6 +55,8 @@ const (
 	PreScaleUpCluster Phase = "pre-scaleup"
 	//PostScaleUpCluster on master0
 	PostScaleUpCluster Phase = "post-scaleup"
+	//UpgradeHost on master0
+	UpgradeCluster Phase = "upgrade"
```
Comment (Starnop) [inline pkg/cluster-runtime/hook.go:59]: mark

File: pkg/cluster-runtime/upgrader.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	if err = registryConfigurator.InstallOn(masters, workers); err != nil {
+		return err
+	}
+
+	if err := i.runClusterHook(master0, UpgradeCluster); err != nil {
+		return err
+	}
+
```
Comment (Starnop) [inline pkg/cluster-runtime/upgrader.go:77]: should we exec cmds after upgrade the cluster

File: pkg/cluster-runtime/upgrader.go [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+	registryConfigurator, err := registry.NewConfigurator(deployHosts, crInfo, i.regConfig, i.infraDriver, i.Distributor)
+	if err != nil {
+		return err
+	}
+
+	if err = registryConfigurator.InstallOn(masters, workers); err != nil {
+		return err
+	}
+
+	if err := i.runClusterHook(master0, UpgradeCluster); err != nil {
```
Comment (kakaZhou719) [inline pkg/cluster-runtime/upgrader.go:74]: pls check if role Or taints still exist after cluster upgrade. use config those info via clusterfile

File: cmd/sealer/cmd/cluster/run.go [hunk 13]
```diff
@@ -115,6 +118,13 @@ func NewRunCmd() *cobra.Command {
 					app, extension, nil, imageEngine, runFlags.Mode)
 			}
 
+			if runFlags.Mode == common.ApplyModeUpgrade {
+				if extension.Type != v12.KubeInstaller {
+					return fmt.Errorf("exit upgrade process, wrong cluster image type: %s", extension.Type)
```
Comment (Starnop) [inline cmd/sealer/cmd/cluster/run.go:123]: wrong cluster image type/wrong image type

**Review by Starnop — APPROVED** [review]
LGTM

**Review by Starnop — APPROVED** [review]
LGTM

**Pull request merged**